sdks/java/io/hadoop-input-format/src/test/java/org/apache/beam/sdk/io/hadoop/inputformat/unit/tests/HadoopInputFormatIOTest.java

```diff
@@ -0,0 +1,796 @@
+    thrown.expectMessage(HadoopInputFormatIOContants.COMPUTESPLITS_EMPTY_SPLITS_ERROR_MSG);
+    parentHIFSource.splitIntoBundles(0, p.getOptions());
+  }
+
+  /**
+   * This test validates behavior of {@link HadoopInputFormatBoundedSource#getEstimatedSizeBytes()
+   * getEstimatedSizeBytes()} and {@link HadoopInputFormatBoundedSource#splitIntoBundles()
+   * splitIntoBundles()} when Hadoop InputFormat's {@link InputFormat#getSplits() getSplits()}
+   * returns NULL value.
+   */
+  @Test
+  public void testSplitIntoBundlesIfGetSplitsReturnsNullValue() throws Exception {
+    HadoopInputFormatBoundedSource<Text, Employee> parentHIFSource = getTestHIFSource(
+            BadNullSplitsInputFormat.class,
+            Text.class,
+            Employee.class,
+            WritableCoder.of(Text.class),
+            AvroCoder.of(Employee.class));
+    thrown.expect(IOException.class);
+    thrown.expectMessage(HadoopInputFormatIOContants.COMPUTESPLITS_NULL_GETSPLITS_ERROR_MSG);
+    parentHIFSource.getEstimatedSizeBytes(p.getOptions());
+
+    thrown.expect(IOException.class);
+    thrown.expectMessage(HadoopInputFormatIOContants.COMPUTESPLITS_NULL_GETSPLITS_ERROR_MSG);
+    parentHIFSource.splitIntoBundles(0, p.getOptions());
+  }
+
+  /**
+   * This test validates behavior of {@link HadoopInputFormatBoundedSource#getEstimatedSizeBytes()
+   * getEstimatedSizeBytes()} and {@link HadoopInputFormatBoundedSource#splitIntoBundles()
+   * splitIntoBundles()} if Hadoop InputFormat's {@link InputFormat#getSplits() getSplits()} returns
+   * InputSplit list having some null values.
+   */
+  @Test
+  public void testSplitIntoBundlesIfGetSplitsReturnsListHavingNullValues() throws Exception {
+    HadoopInputFormatBoundedSource<Text, Employee> hifSource = getTestHIFSource(
+            BadGetSplitsInputFormat.class,
+            Text.class,
+            Employee.class,
+            WritableCoder.of(Text.class),
+            AvroCoder.of(Employee.class));
+    thrown.expect(IOException.class);
+    thrown.expectMessage(HadoopInputFormatIOContants.COMPUTESPLITS_NULL_SPLIT_ERROR_MSG);
+    hifSource.getEstimatedSizeBytes(p.getOptions());
+    thrown.expect(IOException.class);
+    thrown.expectMessage(HadoopInputFormatIOContants.COMPUTESPLITS_NULL_SPLIT_ERROR_MSG);
+    hifSource.splitIntoBundles(0, p.getOptions());
+  }
+
+  /**
+   * This test validates functionality of {@link HadoopInputFormatIO} if user sets wrong key class
+   * and value class.
+   */
+  @Test
+  public void testHIFSourceIfUserSetsWrongKeyOrValueClass() throws Exception {
+    List<BoundedSource<KV<String, String>>> boundedSourceList = getBoundedSourceList(
+            NewObjectsEmployeeInputFormat.class,
+            String.class,
+            String.class,
+            StringUtf8Coder.of(),
+            StringUtf8Coder.of());
+    List<KV<String, String>> bundleRecords = new ArrayList<>();
+    for (BoundedSource<KV<String, String>> source : boundedSourceList) {
```

Comment:
for these, you can just validate that it throws the error for one source, not all of them.